src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java

```diff
@@ -1,26 +1,40 @@
 package pl.cleankod.exchange.provider;
 
-import pl.cleankod.exchange.core.domain.Money;
+import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker;
 import pl.cleankod.exchange.core.gateway.CurrencyConversionService;
 import pl.cleankod.exchange.provider.nbp.ExchangeRatesNbpClient;
 import pl.cleankod.exchange.provider.nbp.model.RateWrapper;
 
 import java.math.BigDecimal;
-import java.math.RoundingMode;
 import java.util.Currency;
 
 public class CurrencyConversionNbpService implements CurrencyConversionService {
     private final ExchangeRatesNbpClient exchangeRatesNbpClient;
+    private final Currency baseCurrency;
+    private final RateWrapper.MidRate baseCurrencyMidRate;
+    private final Currency PL_CURRENCY = Currency.getInstance("PLN");
+    private final String targetNbpTable = "A";
 
-    public CurrencyConversionNbpService(ExchangeRatesNbpClient exchangeRatesNbpClient) {
+    public CurrencyConversionNbpService(ExchangeRatesNbpClient exchangeRatesNbpClient,
+                                        Currency baseCurrency) {
+        this.baseCurrency = baseCurrency;
         this.exchangeRatesNbpClient = exchangeRatesNbpClient;
+        this.baseCurrencyMidRate = new RateWrapper.MidRate(BigDecimal.ONE, baseCurrency);
     }
 
     @Override
-    public Money convert(Money money, Currency targetCurrency) {
-        RateWrapper rateWrapper = exchangeRatesNbpClient.fetch("A", targetCurrency.getCurrencyCode());
-        BigDecimal midRate = rateWrapper.rates().get(0).mid();
-        BigDecimal calculatedRate = money.amount().divide(midRate, RoundingMode.HALF_UP);
-        return new Money(calculatedRate, targetCurrency);
+    @CircuitBreaker(name = "NBP", fallbackMethod = "fallback")
+    public RateWrapper.MidRate getMidRate(Currency targetCurrency) {
+        //this api is really weird, because we can not ask for a default currency (PLN). This is why I made this hacky if below
+        if (baseCurrency == targetCurrency) {
+            return baseCurrencyMidRate;
+        }
+
+        RateWrapper rateWrapper = exchangeRatesNbpClient.fetch(targetNbpTable, targetCurrency.getCurrencyCode());
+        return rateWrapper.getMidRate();
+    }
+
+    private RateWrapper.MidRate fallback(Currency targetCurrency, Throwable e) {
+        return new RateWrapper.MidRate(BigDecimal.ONE, PL_CURRENCY);
```

Comment:
I would ask someone from business what have should be done in case of many errors. I see 2 possible ways:
1. throw an exception 
2. return default mid-rate - in this case mid-rate for PLN